### PR TITLE
Add CatalogSearch no results handle

### PR DESCRIPTION
**Pull request by progreg**
### Description

Added default handle if result count of search results is zero. The feature was implemented for both search variations. For a simple search, the handle will be `catalogsearch_result_index_noresults` and for advanced search, it will be `catalogsearch_advanced_result_noresults`.

### Fixed Issues (if relevant)

1. magento/magento2#18038: Add a custom layout handle when there are no search results

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=18069) <br/>All committers have signed the CLA.

**Comment by magento-engcom-team** [comment]
Hi @progreg. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me $VERSION instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by okorshenko** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @okorshenko. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @okorshenko, here is your [new Magento instance](https://pr-18069.engcom.dev.magento.com).
Admin access: [https://pr-18069.engcom.dev.magento.com/admin](https://pr-18069.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`

**Comment by jakwinkler** [comment]
I'm happy that PR was created to this but... it does not work as expected.
I've implemented this pull request to 2.2.5 version (EE), created a file catalogsearch_result_noresults.xml in Magento CatalogSearch module with following content

```
<?xml version="1.0"?>
<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" layout="1column"
   xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
    <referenceContainer name="header" remove="true" />
    <referenceContainer name="header-wrapper" remove="true" />
    <referenceContainer name="footer" remove="true" />
</page>
```
Am I doing something wrong? Basically the layout xml file is not processed at all... at least thats what I can tell.

For now I will simply create a redirect to a new controller where layout handles work fine.

**Comment by ishakhsuvarov** [comment]
Hi @progreg 
Could you please take a look at the comments above? 

Thanks!

**Comment by progreg** [comment]
@qsolutions-pl, thanks for response. I will check it.

**Comment by progreg** [comment]
Hi @qsolutions-pl, the issue was fixed and now you can check again it by using next handles:

- catalogsearch_result_index_noresults
- catalogsearch_advanced_result_noresults

**Comment by jakwinkler** [comment]
@progreg I do see layout handles being used when I implemented this pull request to my magento installation. Even tests pass when I run them.
But... frontend page still uses 2 columns instead of 1 - that is what I'm trying to accomplish - cache cleared, di cleared :(
To solve this I've created a plugin on Result block and redirect user to a new controller action for now.

**Comment by ishakhsuvarov** [comment]
Hi @qsolutions-pl 
Looks like you already have a solution that is working for you. Feel free to submit it with another Pull Request or even push additional commits to this one.

**Comment by progreg** [comment]
@qsolutions-pl, yeap, you are right. I will take a look again maybe i missed something. Thanks for response

**Comment by progreg** [comment]
@qsolutions-pl, problem with layout was fixed. Please take a look if you still interested in such improvement. Thanks

File: app/code/Magento/CatalogSearch/Test/Unit/Controller/Advanced/ResultTest.php [hunk 1]
```diff
@@ -131,11 +139,66 @@ public function testUrlSetOnException()
...
+            [
+                'context'               => $contextMock,
+                'catalogSearchAdvanced' => $catalogSearchAdvanced,
+                'urlFactory'            => $urlFactoryMock
             ]
         );
         $this->assertEquals($redirectResultMock, $instance->execute());
     }
+
+    public function testNoResultsHandle()
```
Comment (sivaschenko) [inline app/code/Magento/CatalogSearch/Test/Unit/Controller/Advanced/ResultTest.php:166]: please add docblocks/comments for all introduced constants and methods

**Review by sivaschenko — CHANGES_REQUESTED** [review]
@progreg thanks for the contribution! Can you please add phpdocs to all added constants and methods.

File: app/code/Magento/CatalogSearch/Test/Unit/Controller/Advanced/ResultTest.php (hunk 1 above)
Comment (progreg) [reply app/code/Magento/CatalogSearch/Test/Unit/Controller/Advanced/ResultTest.php:166]: done

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-3526](https://jira.corp.magento.com/browse/ENGCOM-3526) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
Hi @progreg. Thank you for your contribution. 
We will aim to release these changes as part of 2.3.1. 
Please check the release notes for final confirmation.

**Pull request merged**